### PR TITLE
Added htmlBodyRenderingOperationWithMessage:folder:rendererDelegate:

### PR DESCRIPTION
**Pull request by zhangao0086**
Added `htmlBodyRenderingOperationWithMessage:folder:rendererDelegate:`.
```objective-c
- (MCOIMAPMessageRenderingOperation *) htmlBodyRenderingOperationWithMessage:(MCOIMAPMessage *)message
																	  folder:(NSString *)folder
															rendererDelegate:(id <MCOHTMLRendererDelegate>)rendererDelegate;
```

### CODE REVIEW

**Comment by zhangao0086** [comment]
@dinhviethoa What do you think?

**Comment by dinhviethoa** [comment]
Sorry, I missed that pull request somehow.

**Comment by zhangao0086** [comment]
Thanks for your Reviews! The inappropriate naming is Fixed!

File: src/async/imap/MCIMAPMessageRenderingOperation.cpp [hunk 1]
```diff
@@ -55,22 +67,27 @@ void IMAPMessageRenderingOperation::main()
...
-        mResult = session()->session()->plainTextRendering(mMessage, folder(), &error);
-    }
-    else if (mRenderingType == IMAPMessageRenderingTypePlainTextBody) {
-        mResult = session()->session()->plainTextBodyRendering(mMessage, folder(), false, &error);
-    }
-    else if (mRenderingType == IMAPMessageRenderingTypePlainTextBodyAndStripWhitespace) {
-        mResult = session()->session()->plainTextBodyRendering(mMessage, folder(), true, &error);
-    }
-    
+	if (mHtmlCallBack != NULL) {
```
Comment (dinhviethoa) [inline src/async/imap/MCIMAPMessageRenderingOperation.cpp]: Could you fix the indentation here?

File: src/async/imap/MCIMAPMessageRenderingOperation.h [hunk 2]
```diff
@@ -24,7 +24,10 @@ namespace mailcore {
         
         virtual void setRenderingType(IMAPMessageRenderingType type);
         virtual IMAPMessageRenderingType renderingType();
-        
+		
+		virtual void setHtmlCallBack(HTMLRendererTemplateCallback * htmlCallBack);
```
Comment (dinhviethoa) [inline src/async/imap/MCIMAPMessageRenderingOperation.h]: here too.

File: src/async/imap/MCIMAPMessageRenderingOperation.h [hunk 3]
```diff
@@ -38,6 +41,7 @@ namespace mailcore {
         IMAPMessageRenderingType mRenderingType;
         String * mResult;
         IMAPMessage * mMessage;
+		HTMLRendererTemplateCallback * mHtmlCallBack;
```
Comment (dinhviethoa) [inline src/async/imap/MCIMAPMessageRenderingOperation.h]: and here.

File: src/core/imap/MCIMAPSession.h [hunk 4]
```diff
@@ -203,6 +203,8 @@ namespace mailcore {
         
         /** HTML rendering of the body of the message.*/
         virtual String * htmlBodyRendering(IMAPMessage * message, String * folder, ErrorCode * pError);
+		
+		virtual String * htmlBodyRendering(IMAPMessage * message, String * folder, HTMLRendererTemplateCallback * htmlCallback, ErrorCode * pError);
```
Comment (dinhviethoa) [inline src/core/imap/MCIMAPSession.h]: indentation.

**Comment by zhangao0086** [comment]
I think it's `Files change`'s bug. The following is the part of the original code:
```c++
namespace mailcore {
    
    class MAILCORE_EXPORT IMAPMessageRenderingOperation : public IMAPOperation {
    public:
        IMAPMessageRenderingOperation();
        virtual ~IMAPMessageRenderingOperation();
        
        virtual void setRenderingType(IMAPMessageRenderingType type);
        virtual IMAPMessageRenderingType renderingType();
		
		virtual void setHtmlCallBack(HTMLRendererTemplateCallback * htmlCallBack);
		virtual HTMLRendererTemplateCallback * htmlCallBack();
		
        virtual void setMessage(IMAPMessage * message);
        virtual IMAPMessage * message();
        
        // Result.
        virtual String * result();
        
    public: // subclass behavior
        virtual void main();
    
    private:
        IMAPMessageRenderingType mRenderingType;
        String * mResult;
        IMAPMessage * mMessage;
		HTMLRendererTemplateCallback * mHtmlCallBack;
        
    };
    
}
```

File: src/objc/abstract/MCOAbstractMessageRendererCallback.mm [hunk 5]
```diff
@@ -11,17 +11,34 @@
...
+#import <Objc/runtime.h>
 
 using namespace mailcore;
 
 MCOAbstractMessageRendererCallback::MCOAbstractMessageRendererCallback(MCOAbstractMessage * message, id <MCOHTMLRendererDelegate> rendererDelegate,
                                    id <MCOHTMLRendererIMAPDelegate> rendererIMAPDelegate)
 {
     mMessage = message;
-    mRendererDelegate = rendererDelegate;
+	if (rendererDelegate) {
```
Comment (dinhviethoa) [inline src/objc/abstract/MCOAbstractMessageRendererCallback.mm]: What is this change for?

File: src/core/imap/MCIMAPSession.cpp [hunk 6]
```diff
@@ -4359,6 +4359,25 @@ String * IMAPSession::htmlBodyRendering(IMAPMessage * message, String * folder,
...
+	HTMLRendererIMAPDataCallback * dataCallback = new HTMLRendererIMAPDataCallback(this, message->uid());
+	String * htmlBodyString = HTMLRenderer::htmlForIMAPMessage(folder,
+															   message,
+															   dataCallback,
+															   htmlCallback);
+	
+	* pError = dataCallback->error();
+	
+	if (* pError != ErrorNone) {
+		return NULL;
```
Comment (dinhviethoa) [inline src/core/imap/MCIMAPSession.cpp]: release dataCallback properly here.

Comment (zhangao0086) [inline src/core/imap/MCIMAPSession.cpp]: Fixed!

File: src/objc/abstract/MCOAbstractMessageRendererCallback.mm (hunk 5 above)
Comment (zhangao0086) [inline src/objc/abstract/MCOAbstractMessageRendererCallback.mm]: I think the `MCOAbstractMessageRendererCallback` should retain the `rendererDelegate`.
Consider the following code:
```objectivec
id<MCOHTMLRendererDelegate> render = [CustomRenderer new];
MCOIMAPMessageRenderingOperation *renderingOperation = [self.IMAPSession htmlBodyRenderingOperationWithMessage:rawMessage
																										folder:folder
																							  rendererDelegate:render];
[renderingOperation start:^(NSString *HTMLString, NSError *error) {
	BLOCK_SAFE_CALLS(completeBlock, HTMLString, error);
}];
```
and this:
```objectivec
- (MCOIMAPMessageRenderingOperation *) htmlBodyRenderingOperationWithMessage:(MCOIMAPMessage *)message
																	  folder:(NSString *)folder
															rendererDelegate:(id<MCOHTMLRendererDelegate>)rendererDelegate
{
	MCOAbstractMessageRendererCallback * htmlCallBack = new MCOAbstractMessageRendererCallback(NULL, rendererDelegate, NULL);
	IMAPMessageRenderingOperation * coreOp = MCO_NATIVE_INSTANCE->htmlRenderingOperation(MCO_FROM_OBJC(IMAPMessage, message), [folder mco_mcString], htmlCallBack);
	htmlCallBack->release();
	return MCO_TO_OBJC(coreOp);
}
```

and this:
```objectivec
void OperationQueue::startThread()
{
    if (mStarted)
        return;
    
    if (mCallback) {
        mCallback->queueStartRunning();
    }
    
    retain(); // (3)
    mQuitting = false;
    mStarted = true;
    pthread_create(&mThreadID, NULL, (void * (*)(void *)) OperationQueue::runOperationsOnThread, this);
    mailsem_down(mStartSem);
}
```

if the `MCOAbstractMessageRendererCallback` do not retain the `rendererDelegate`, the `mRendererDelegate` will be released before `IMAPMessageRenderingOperation::main`.

File: src/async/imap/MCIMAPMessageRenderingOperation.cpp (hunk 1 above)
Comment (dinhviethoa) [inline src/async/imap/MCIMAPMessageRenderingOperation.cpp]: You need to use spaces and not tabs

File: src/objc/abstract/MCOAbstractMessageRendererCallback.mm [hunk 7]
```diff
@@ -11,17 +11,34 @@
 #import "MCOHTMLRendererDelegate.h"
 #import "MCOHTMLRendererIMAPDelegate.h"
 #import "MCOUtils.h"
+#import <Objc/runtime.h>
```
Comment (dinhviethoa) [inline src/objc/abstract/MCOAbstractMessageRendererCallback.mm]: This line is a big issue in the architecture. The C++ part is compatible with more than just Objective-C platforms.

**Comment by dinhviethoa** [comment]
Ok. I feel a little bit bad here to analyzing this code properly earlier.
My thought is that if you need to achieve the same effect with safer code, you can use:

```
- (NSString *) htmlRenderingWithFolder:(NSString *)folder
                              delegate:(id <MCOHTMLRendererIMAPDelegate>)delegate;
```

Please take a look at the example to see how to handle it.

The issue with the code you're showing is that the API of `id<MCOHTMLRendererDelegate>` would accessed in a thread, while you created it on the main thread.
It would break some assumptions of ObjC conventions.

**Comment by zhangao0086** [comment]
I've tried to use `htmlRenderingWithFolder:delegate:`, but it gave me a nil.
Here is my code:
```objectivec
MCOIndexSet *uids = [MCOIndexSet indexSetWithIndex:4318];

MCOIMAPMessagesRequestKind requestKind = MCOIMAPMessagesRequestKindUid | MCOIMAPMessagesRequestKindFullHeaders | MCOIMAPMessagesRequestKindFlags | MCOIMAPMessagesRequestKindStructure | MCOIMAPMessagesRequestKindInternalDate;
MCOIMAPFetchMessagesOperation *operation = [self.IMAPSession fetchMessagesOperationWithFolder:folder
                                                                                  requestKind:requestKind
                                                                                         uids:uids];
[operation start:^(NSError * _Nullable error, NSArray * _Nullable messages, MCOIndexSet * _Nullable vanishedMessages) {
    MCOIMAPMessage *message = messages.firstObject;
    NSString *HTMLString = [message htmlRenderingWithFolder:folder delegate:[WBMailHTMLRenderer new]];
    NSLog(@"%@", HTMLString);
}];
```

Here is log:
```
Mail Log ========== Type(0)
* 4302 FETCH (UID 4318 FLAGS () INTERNALDATE "24-Jul-2016 08:43:39 +0800" BODYSTRUCTURE ("text" "html" ("charset" "utf-8") NIL NIL "quoted-printable" 22581 309 NIL NIL NIL) BODY[HEADER.FIELDS (FROM TO CC SUBJECT DATE MESSAGE-ID REFERENCES IN-REPLY-TO)] {340}

From: "337262953@qq.com" <337262953@qq.com>

To: zhangao <zhangao@shaozi.com>

Subject: =?utf-8?B?5p6X6aOeKDk1MjU1NDU0KeW6lOiBmO+8mklPUw==?= =?utf-8?B?56CU5Y+R5bel56iL5biI77yI5Yid57qn77yJ?= =?utf-8?B?LWZyb23ljqbpl6jkurrmiY3nvZE=?=

Date: Sun, 24 Jul 2016 08:43:10 +0800

Message-ID: <552974631.3537314.1469320990125.JavaMail.root@vt153>



)
```

I guess because `fetchMessagesOperationWithFolder:requestKind:uids:` getting only a small amount of data:`FROM TO CC SUBJECT DATE MESSAGE-ID REFERENCES IN-REPLY-TO`, it's not enough information to generate the HTML string.

I'll let you know if I find a better way...

**Comment by zhangao0086** [comment]
Is there a way to use `IMAPFetchParsedContentOperation` only fetch `mainPart` data?

**Comment by dinhviethoa** [comment]
`htmlRenderingWithFolder:delegate:` is going to call you delegate if it needs data.
Then, you need to fetch the data (using fetchPart).
And then, you call again `htmlRenderingWithFolder:delegate:` when you have everything available.

**Comment by zhangao0086** [comment]
Can I use `requiredPartsForRendering` like this:

```objective-c
NSArray<MCOIMAPPart *> *partsForRendering = [rawMessage requiredPartsForRendering];
if (partsForRendering.count > 0) {
    [partDownloadManager addObserverForMailId:mail.mailId completeBlock:^(NSError *error) {
        // All parts have been downloaded.
        if (error == nil) {
            WBMailHTMLRenderer2 *renderer = [WBMailHTMLRenderer2 new];
            [renderer setFetchPart:^NSData *(MCOIMAPPart *part) {
                return [partDownloadManager partData:part mail:mail];
            }];
            NSString *HTMLString = [rawMessage htmlRenderingWithFolder:folder delegate:renderer];
            // NSLog(@"%@", HTMLString);
        } else {
            // has error
        }
    }];
    
    for (MCOIMAPPart *part in partsForRendering) {
    	// start downloading part
    }
} else {
	// no parts for rendering
}
```

**Comment by dinhviethoa** [comment]
Sure! That's a possible implementation.

**Comment by zhangao0086** [comment]
Thanks! Closing this thread...

**Pull request closed without merging**